torch/distributed/distributed_c10d.py

```diff
@@ -1131,6 +1135,12 @@ def all_gather(tensor_list,
         tensor (Tensor): Tensor to be broadcast from current process.
         group (ProcessGroup, optional): The process group to work on
         async_op (bool, optional): Whether this op should be an async op
+        no_copy (bool, optional): Do not unflatten output tensors if they
+            are already contiguous views into an already flattened tensor
+            and backend is NCCL. If input tensor shares storage with
+            output tensors, it must be properly aligned, i.e.
+            offset == rank * flat_output.numel() // world_size.
```

Comment:
Could you be a bit more specific here, e.g. something like
"""
all input and output tensors must be in the same storage and properly aligned, i.e. offset(input tensor in i-th rank) = i * num_elements(input tensor)
"""

Reply:
Hi, there are actually two modes here. The performance benefit we are after here is gained by not having to unflatten the outputs if the output tensors are views into an already flattened tensor. The input tensor does not have to share storage with the outputs, but this is allowed and can save some memory. If input and outputs does share storage, input has to be properly aligned, that's what I was trying to convey with that paragraph. I'll try to rephrase it to make it clearer.